pkg/operator/ceph/object/spec_test.go

```diff
@@ -876,3 +877,160 @@ func TestAWSServerSideEncryption(t *testing.T) {
+		{"dns name with zone with custom endpoints ceph v19", []string{"my.dns.name1", "my.dns.name2"}, "--rgw-dns-name=my.dns.name1,my.dns.name2,rook-ceph-rgw-default.mycluster.svc,my.custom.endpoint1,MY.CUSTOM.ENDPOINT2", cephver.CephVersion{Major: 19, Minor: 0, Extra: 0}, "myzone", []string{"http://my.custom.endpoint1:80", "http://MY.CUSTOM.ENDPOINT2:80"}, false},
+		{"dns name with zone with custom invalid endpoints ceph v19", []string{"my.dns.name1", "my.dns.name2"}, "", cephver.CephVersion{Major: 19, Minor: 0, Extra: 0}, "myzone", []string{"http://my.custom.endpoint:80", "http://!my.custom.invalid-endpoint:80"}, true},
+		{"dns with zone with mixed invalid and valid dnsnames/custom endpoint ceph v19", []string{"my.dns.name", "!my.invalid-dns.name"}, "", cephver.CephVersion{Major: 19, Minor: 0, Extra: 0}, "myzone", []string{"http://my.custom.endpoint1:80", "http://!my.custom.invalid-endpoint:80"}, true},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := setupTest(tt.zoneName, tt.cephvers, tt.dnsNames, tt.CustomEndpoints)
+			res, err := c.addDNSNamesToRGWServer()
+			if tt.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+			assert.Equal(t, tt.expectedDNSArg, res)
+
+		})
+	}
+}
+
+func TestGetHostnameFromEndpoint(t *testing.T) {
+	tests := []struct {
+		name     string
+		endpoint string
+		expected string
+		wantErr  bool
+	}{
+		{"empty endpoint", "", "", true},
+		{"http endpoint", "http://my.custom.endpoint1:80", "my.custom.endpoint1", false},
+		{"https endpoint", "https://my.custom.endpoint1:80", "my.custom.endpoint1", false},
+		{"http endpoint without port", "http://my.custom.endpoint1", "my.custom.endpoint1", false},
+		{"https endpoint without port", "https://my.custom.endpoint1", "my.custom.endpoint1", false},
+		{"multiple portocol endpoint case 1", "http://https://my.custom.endpoint1:80", "", true},
+		{"multiple portocol endpoint case 2", "https://http://my.custom.endpoint1:80", "", true},
+		{"ftp endpoint", "ftp://my.custom.endpoint1:80", "my.custom.endpoint1", false},
+		{"custom protocol endpoint", "custom://my.custom.endpoint1:80", "my.custom.endpoint1", false},
+		{"custom protocol endpoint without port", "custom://my.custom.endpoint1", "my.custom.endpoint1", false},
+		{"invalid endpoint", "http://!my.custom.endpoint1:80", "", true},
+		{"invalid endpoint multiple ports", "http://my.custom.endpoint1:80:80", "", true},
+		{"http endpoint with upper case", "http://MY.CUSTOM.ENDPOINT1:80", "MY.CUSTOM.ENDPOINT1", false},
+		{"http endpoint with upper case without port", "http://MY.CUSTOM.ENDPOINT1", "MY.CUSTOM.ENDPOINT1", false},
+		{"https endpoint with upper case", "https://MY.CUSTOM.ENDPOINT1:80", "MY.CUSTOM.ENDPOINT1", false},
+		{"https endpoint with upper case without port", "https://MY.CUSTOM.ENDPOINT1", "MY.CUSTOM.ENDPOINT1", false},
```

Comment:
I'm pretty sure that it is best if we convert these to lower case. Have you had time to test this with the wildcard ingress to see if upper-case letters in hostnames cause errors?

Reply:
* spec.rules[0].host: Invalid value: "RGW.example": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
* spec.rules[1].host: Invalid value: "*.RGW.example": a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\*\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

Ingress follows the  RFC 1123, so I am guessing it is better treat capital letter hostname as invalid case for our usecase